### DISCUSSION TITLE
How to set Map configuration properties using Environment Variables ?

### DISCUSSION THREAD

**Discussion by amoscatelli** [quarkusio/quarkus#30353] [Q&A] [answered]
Hi everybody.

There are several Map configuration properties and, according to documentation (https://quarkus.io/guides/all-config), they can be set using Environment Variables. 

I am trying to use QUARKUS_KUBERNETES_ENV_VARS.

Yet, setting an environment variable like this : 

`QUARKUS_KUBERNETES_ENV_VARS_TESTNAME=testvalue`

doesn't seem to have any effect on my kubernates.yml.
This happens with both quarkus:dev Maven goal, or the quarkus:build Maven goal (using JIB).

At last, if I set (inside application.properties)

`quarkus.kubernetes.env.vars.testname=testvalue`

then the configuration is honored and I can see the environment variable inside my kubernates.yml.

This seems to happen with other Map configs, for example : 

quarkus.container-image.labels-labels / QUARKUS_CONTAINER_IMAGE_LABELS

Other (non Map) environment variables are correctly working.

Can somebody confirm this is supposed to work ?
Please paste ANY example of environment variable configuration for any Map property, I could find **none**.

I already have different workarounds in mind, but I'd rather have this work just like this, so please help.

Thank you in advance !

[a comment by an automated account is not shown]

**Comment by geoand** [comment, marked as answer]
@radcortez can you shed some light on this? Perhaps we can also add an example to the config docs?

**Reply by radcortez** [reply]
The rules for Environment Variables are documented here:
https://quarkus.io/guides/config-reference#environment-variables

SmallRye Config defines some additional rules (linked from the Quarkus main docs):
https://smallrye.io/smallrye-config/latest/config/environment-variables/

But this has been a recurring question, so we need to detail the rules directly in the Quarkus docs. I can do that.

Regarding the Kubernetes Config, we have an issue mapping environment variables and system properties for build time config and a `Map`. As you know, we rely on the list of property names to populate a `Map`. For buil time, we override the System and Env sources to return an empty list since we don't want to record variables from the build environment. It is one-way only; we cannot map them back on runtime. 

This has been present in Quarkus for a long time, but it seems we never noticed it. I'll think of a way to fix this.

**Reply by geoand** [reply]
> But this has been a recurring question, so we need to detail the rules directly in the Quarkus docs. I can do that.

Great thanks! I think it will help a lot as for example I never remember where to look either.

> This has been present in Quarkus for a long time, but it seems we never noticed it. I'll think of a way to fix this.

Thanks

**Reply by amoscatelli** [reply]
> The rules for Environment Variables are documented here: https://quarkus.io/guides/config-reference#environment-variables
> 
> SmallRye Config defines some additional rules (linked from the Quarkus main docs): https://smallrye.io/smallrye-config/latest/config/environment-variables/
> 
> But this has been a recurring question, so we need to detail the rules directly in the Quarkus docs. I can do that.
> 
> Regarding the Kubernetes Config, we have an issue mapping environment variables and system properties for build time config and a `Map`. As you know, we rely on the list of property names to populate a `Map`. For buil time, we override the System and Env sources to return an empty list since we don't want to record variables from the build environment. It is one-way only; we cannot map them back on runtime.
> 
> This has been present in Quarkus for a long time, but it seems we never noticed it. I'll think of a way to fix this.

@radcortez 

Yea I had already consulted the SmallRye additional rules (your link is broken with a 404 btw), but nothing there solved the issue.

I am not sure to understand the technical detail you are talking about.
Do you want me to file an issue and link this discussion ?

Is this specific for the Kubernates plugin/configuration or is related to every Map property ?

Can you confirm the syntax I am using is indeed correct ?
`QUARKUS_KUBERNETES_ENV_VARS_TESTNAME=testvalue`

p.s. correct link for SmallRye Config Environment Variables mapping : 
https://smallrye.io/smallrye-config/3.1.1/config/environment-variables/

Thank you for your support

**Reply by radcortez** [reply]
Yes, sorry. Copy the wrong link. Thanks for looking for the right one.

The env var name is correct; it is just that there is a bug on our side for env vars set in the config build phase for every Map property. It works for runtime config. I'm working on a fix.

**Reply by amoscatelli** [reply]
Ok thank you !

**Reply by radcortez** [reply]
This should fix it: https://github.com/quarkusio/quarkus/pull/30368

**Reply by radcortez** [reply]
And docs: https://github.com/quarkusio/quarkus/pull/30369

**Reply by amoscatelli** [reply]
Thank you again.
I am using Quarkus 3.0.0.Alpha2, will the fix available on the next (alpha) 3.x release ?

**Reply by radcortez** [reply]
Not sure, but probably.

**Reply by amoscatelli** [reply]
@radcortez ummn I can't see the 30368 among the fixes released with Alpha4 : 
https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha4

Strange thing is that I can see the 30369

Is there a reason for this ?
Thank you in advance

**Reply by amoscatelli** [reply]
Yea I confirm this still doesn't work for Alpha4 :(

**Reply by radcortez** [reply]
I've tried a sample project with Alpha4, and your initial description, and it worked for me. Can you provide me a reproducer, the repro steps, and the expected result?

**Reply by amoscatelli** [reply]
You are right.

QUARKUS_KUBERNETES_ENV_VARS_TESTNAME works

But QUARKUS_KUBERNETES_ENV_VARS_MP_JWT_VERIFY_ISSUER still doesn't work

What do I have to do to use keys containing the "_" character ?

**Reply by amoscatelli** [reply]
Seems like the correct sytax is : 
QUARKUS_KUBERNETES_ENV_VARS__MP_JWT_VERIFY_ISSUER__

**Reply by amoscatelli** [reply]
Maybe we should have an example in the doc about this

**Reply by radcortez** [reply]
Can you open a PR with it? Thank you :)

**Reply by amoscatelli** [reply]
Sorry but I can't.
The reason is  I don't know why it works with this syntax.

QUARKUS_KUBERNETES_ENV_VARS__MP_JWT_VERIFY_ISSUER__ makes no sense since every _ should be a special character
I can't really explain the last __ to myself

I just found this by trial and error.

I think that some guy from microprofile config should explain this ...
I even tried to debug but I didn't solve the mistery

**Reply by amoscatelli** [reply]
And to tell the truth I don't really know if this is really a microprofile config thing, or this is some special rule/conversion coming from quarkus config extension

**Reply by radcortez** [reply]
Quarkus and SmallRye Config is well ahead of MP Config :) SmallRye Config defines additional rules to handle environment variables.

This is somehow a special case, because this Map contains composed keys to other configuration names. The issue is that we don't know if a composed dotted key is going to be used to map another property inside the configuration group or if it is part of the key itself. As an example:

`quarkus.kubernetes.env.vars.var=value`-> `var=value` -> This is ok because the key is a single name
`quarkus.kubernetes.env.vars.my-env-var=value` -> `my-env-var=value` -> Again, ok because the key is a single name
`quarkus.kubernetes.env.vars.my.env.var=value` -> EMPTY -> This is a composed key, but we don't know which part is the actual key an we can't assume
`quarkus.kubernetes.env.vars."my.env.var"=value` -> `my.env.var=value` -> Using quotes help us understand that `my.env.var` is the entire key, so it works.

In the case of Environment variables a double `__` represents the `."`, so that is why you to surround the key with `__` in an environment variable.

**Reply by amoscatelli** [reply]
ok thank you for the explanation !

**Comment by amoscatelli** [comment]
@radcortez ?

**Reply by radcortez** [reply]
Sorry, I was on PTO, but I'm glad you were able to figure it out :)
